workload : fix tpcc multiple workload load generation

csgourav opened the pull request:
This change addresses three things for multiple workload load generation for tpcc workload for tpcc benchmarking
1. Partitions assignment to warehouses based on affinity partitions and active warehouses
    - Earlier partition assignment to workers didn't take into account affinity partition, this change takes care of that
    - Also, would skip assigning some partitions to some worker, as the isMyPart() checked happened after partition assignment, we fix this by taking into account affinity partitions early
2. Updating audit checks based on count of affinity partitions
   - Now that we are generating different load from multiple workload nodes, the audit checks needs to be adjusted to take this into account, as there is no aggregated audit checks for multiple workload nodes.
   - Introduced partitionFactor to take into account number of workload nodes, and decrease the overall numbers for audit checks by this factor.
3. Correcting partitions to connections urls mapping
     - URL mapping happens for each partition based on the urls passed in the workload command. This fix makes sure we do url to partition mapping correctly taking into account the affinity partitions. Earlier we just ignored affinity partitions in this part.

Epic: none
Release: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/134796)
<!-- Reviewable:end -->

sambhav-jain-16 commented on pkg/workload/tpcc/tpcc.go:
```diff
@@ -1014,13 +1028,21 @@ func (w *tpcc) Ops(
 		if len(w.multiRegionCfg.regions) == 0 {
-			warehouse = w.wPart.totalElems[workerIdx%len(w.wPart.totalElems)]
-			p = w.wPart.partElemsMap[warehouse]
+			for {
+				elemIndex = elemIndex + 1
+				warehouse = w.wPart.totalElems[elemIndex%len(w.wPart.totalElems)]
+				part := w.wPart.partElemsMap[warehouse]
+				if isMyPart(part) {
+					p = part
+					break
+				}
+			}
```
1. Is this needed: `elemIndex%len(w.wPart.totalElems)`? Is this because `totalElems` is changed by a different go routine?

2. Can we make the for loop better with maybe something like
```
part := -1
for !isMyPart(part) {
        elemIndex = elemIndex + 1
				warehouse = w.wPart.totalElems[elemIndex%len(w.wPart.totalElems)]
				part := w.wPart.partElemsMap[warehouse]
}

srosenberg replied:
For better readability, I'd replace the inner loop with a helper; e.g.,

```
warehouse, p = findNextPartition(elemIndex, w.wPart.totalElems, w.wPart.partElemsMap, isMyPart)
```

where 

```
func findNextPartition(startIndex int, totalElems []int, partElemsMap map[int]int, isMyPart func(int) bool) (int, int) {
	for {
		startIndex++
		warehouse := totalElems[startIndex%len(totalElems)]
		p := partElemsMap[warehouse]
		if isMyPart(p) {
			return warehouse, p
		}
	}
return -1, -1
}
```

csgourav replied:
Done refactored, added function.

sambhav-jain-16 reviewed (commented):
Thanks @csgourav for the PR
Some comments and questions

srosenberg commented on pkg/workload/tpcc/audit.go:
```diff
@@ -22,7 +22,16 @@ const (
 
-	warehouses int
+	warehouses         int
+	affinityPartitions []int
```
Nit: `partitionAffinity` since "affinity" is a property we want to ascribe to a given partition. We also typically drop plural for arrays.

srosenberg replied:
I didn't realize it was an existing flag; funny enough, the flag is `--partition-affinity`.

csgourav replied:
Done updated.

srosenberg commented on pkg/workload/tpcc/audit.go:
```diff
@@ -212,15 +231,26 @@ func check92253(a *auditor) auditResult {
+		for _, partition := range a.affinityPartitions {
+			for _, warehouse := range a.part.partElems[partition] {
+				if _, ok := a.orderLineRemoteWarehouseFreq[warehouse]; !ok {
+					return newFailResult("no remote order-lines for warehouses %d", warehouse)
```
You could use a helper func to return a slice of warehouses from `affinityPartitions`, and get rid of the duplicated for-loop.

csgourav replied:
Done updated.

srosenberg commented on pkg/workload/tpcc/audit.go:
```diff
@@ -250,12 +280,22 @@ func check92254(a *auditor) auditResult {
 	}
 
-	if remotePayments < 15*uint64(a.warehouses) {
+	if remotePayments < 15*uint64(a.warehouses/a.partitionFactor) {
```
Would this violate the spec. if partition sizes are not exactly the same?

csgourav replied:
I think, this could give false negative, but not false positive. Since we kind of round down the partition factor.
Which should be okay I think, and the difference from actual would not be much.

srosenberg commented on pkg/workload/tpcc/tpcc.go:
```diff
@@ -965,9 +973,15 @@ func (w *tpcc) Ops(
-			p := i % w.partitions
-			partitionDBs[p] = append(partitionDBs[p], db)
+		if len(w.affinityPartitions) == 0 {
+			for i, db := range dbs {
```
This can be simplified to,

```
for i, db := range dbs {
	var p int
	if len(w.affinityPartitions) == 0 {
		p = i % w.partitions
	} else {
		p = w.affinityPartitions[i]
	}
	partitionDBs[p] = append(partitionDBs[p], db)
}
```

With branch prediction, there isn't really a penalty for moving the conditional inside.

csgourav replied:
Since, affinityPartitions will usually be less than dbs length, this would cause error I think.

csgourav commented on pkg/workload/tpcc/tpcc.go:
```diff
@@ -594,14 +596,20 @@ func (w *tpcc) Hooks() workload.Hooks {
 			const totalHeader = "\n_elapsed_______tpmC____efc__avg(ms)__p50(ms)__p90(ms)__p95(ms)__p99(ms)_pMax(ms)"
 			fmt.Println(totalHeader)
 
+			partitionFactor := 1
```
Done.

csgourav commented on pkg/workload/tpcc/audit.go:
```diff
@@ -250,12 +280,22 @@ func check92254(a *auditor) auditResult {
+	} else {
+		for _, i := range a.affinityPartitions {
+			for _, p := range a.part.partElems[i] {
+				if _, ok := a.paymentRemoteWarehouseFreq[p]; !ok {
```
Done

csgourav committed:
workload: fix tpcc multiple workload load generation

This change addresses three things for multiple workload load generation for tpcc workload

1. Fixes audit checks based on affinity partitions, takes into account multiple workload nodes
2. Partitions assignment to worker incase of affinity partitions
3. dburls assignment to partitions becoming affinity partitions aware

Epic: none
Release: none

csgourav commented:
TFTR

csgourav commented:
bors r+
